extension/ts/copy.ts

```diff
@@ -0,0 +1,90 @@
+import { RELEVANT_TAGS } from "./constants";
+import { SaveData } from "./types";
+import { getElementsByTags } from "./util";
+
+const COLLECTIONS_ID_PREFIX = "collection_u_";
+const SAVE_DATA_KEY = "_save-data";
+const COLLECTIONS_KEY = "___collections_";
+
+const saveData = (parent: HTMLElement) => (event: Event) => {
+	event.preventDefault();
+	localStorage.setItem(SAVE_DATA_KEY, JSON.stringify(getSaveData(parent)));
+};
+
+const loadData = (parent: HTMLElement) => (event: Event) => {
+	event.preventDefault();
+	const saveData = JSON.parse(localStorage.getItem(SAVE_DATA_KEY) ?? "{}");
+	insertSaveData(parent, saveData);
+};
+
+const getSaveData = (parent: HTMLElement) => {
+	const elements = getElementsByTags(parent, RELEVANT_TAGS);
+	return elements.reduce((saveData: SaveData, element: any) => {
+		// We can't change hidden elements because LibraryThing relies
+		// on hidden form inputs to send additional, form-specific metadata
+		// on save
+		if (element && element.id && element.type !== "hidden") {
+			const { value, checked } = element;
+			if (element.id.startsWith(COLLECTIONS_ID_PREFIX)) {
+				const collections = saveData[COLLECTIONS_KEY] || {};
+				const [span] =
+					element.parentElement.getElementsByTagName("span");
+				collections[span.textContent] = { value, checked };
+				saveData[COLLECTIONS_KEY] = collections;
+			} else {
+				saveData[element.id] = { value, checked };
+			}
+		}
+		return saveData;
+	}, {});
+};
+
+const insertSaveData = (parent: HTMLElement, saveData: SaveData) => {
+	const elements = getElementsByTags(parent, RELEVANT_TAGS);
+	return elements.forEach((element: any) => {
+		// We can't change hidden elements because LibraryThing relies
+		// on hidden form inputs to send additional, form-specific metadata
+		// on save
+		if (element && element.id && element.type !== "hidden") {
+			let saveElement = element;
+			if (element.id.startsWith(COLLECTIONS_ID_PREFIX)) {
+				const span =
+					element.parentElement.getElementsByTagName("span")[0];
+				saveElement =
+					saveData[COLLECTIONS_KEY][span.textContent] || element;
```

Comment:
super nit: this probably always exists, but for the back burner maybe can change to `element.parentElement.getElementsByTagName("span")?.[0]` and `span?.textContent`

Reply:
`getElementsByTagName` always will give you back a collection, but `span` could be `undefined`. So yeah you'd want the optional chaining on `span.textContent`. I think that should be in another PR for generally making the code safer.